### PR TITLE
[CoSim] Conversion Utilities

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
In CoSimulation problems, there are some situations in which variables or quantities (say FORCE) received from other solvers (say OpenFOAM) on a **Element** location and if we want it to be on a **Nodal** location (say for KRATOS point of view), then in this situation we need to use small conversion techniques. Hence this is PR is created for Conversion Utilities (currently only for FORCE).
 
- CoSimulationApplication
- ConversionUtilities

**Changelog**
- Added ConversionUtilities Class.
- Implemented a function to convert an elemental Force data to nodal data.
- Added cpp test for this implementation.
- Created a python binding of this functionality.
- Added **ElementalToNodalData** operation in the existing Coupling operations (for CoSimulation App).
- Added test for **ElementalToNodalData** operation.

### CODE REVIEW

File: applications/CoSimulationApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 1]
```diff
@@ -54,6 +55,12 @@ namespace Python{
             .value("Origin", FetiSolverIndexType::Origin)
             .value("Destination", FetiSolverIndexType::Destination)
             ;
+
+        pybind11::class_< ConversionUtilities>(m, "ConversionUtilities")
+            .def("ConvertElementalDataToNodalData",
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_python/add_custom_utilities_to_python.cpp]:
```suggestion
            .def_static("ConvertElementalDataToNodalData",
```

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Author1 Fullname
+//                   Author2 Fullname
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]: you can put our names

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "includes/variables.h"
+#include "includes/model_part.h"
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]: this should be in the header

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#include "utilities/parallel_utilities.h"
+#include "utilities/variable_utils.h"
+#include "utilities/atomic_utilities.h"
+#include "conversion_utilities.h"
+
+namespace Kratos {
+
+void ConversionUtilities::ConvertElementalDataToNodalData(ModelPart& model_part_interface)
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]:
```suggestion
void ConversionUtilities::ConvertElementalDataToNodalData(ModelPart& rModelPart)
```

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.h [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    static void ConvertElementalDataToNodalData(ModelPart& model_part_interface);
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.h]:
```suggestion
    static void ConvertElementalDataToNodalData(ModelPart& rModelPart);
```

File: applications/CoSimulationApplication/tests/cpp_tests/test_conversion_utilities.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        auto nodal_force = r_node.FastGetSolutionStepValue(FORCE);
+        for(int i = 0 ; i< 3; i++){
+            nodal_force_values[counter++] = nodal_force[i];
+        }
+    }
+
+    // Expected Values at Nodes
+    array_1d<double, 18> expected_values {3, 2 ,0, 6, 4, 0, 6, 4 ,0, 3, 2 ,0, 3, 2 ,0, 3, 2 ,0};
+
+    KRATOS_CHECK_VECTOR_NEAR(nodal_force_values, expected_values, 1.0e-4)
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/cpp_tests/test_conversion_utilities.cpp]: why such a large tolerance?
this should be quite tight

File: applications/CoSimulationApplication/tests/test_coupling_operations.py [hunk 7]
```diff
@@ -141,6 +141,71 @@ def __CompareValues(self, old_data, new_data, factor):
...
+            "variable_name"         : "FORCE",
+            "location"              : "node_historical",
+            "dimension"             : 3
+        }""")
+
+        nodal_data_output = CouplingInterfaceData(nodal_data_output_setting,  self.model)
+
+        expected_nodal_values = [3, 2 ,0, 7, 3, 0, 7, 3 ,0, 3, 2 ,0, 4, 1 ,0, 4, 1 ,0 ]
+
+        self.__CheckData(expected_nodal_values, nodal_data_output.GetData())
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/test_coupling_operations.py]:
```suggestion
        self.assertVectorEqual(expected_nodal_values, nodal_data_output.GetData())
```
this should also work

**Review by philbucher — COMMENTED** [review]
looks good, minor comments, thx!

and please revert the changes related to the predictors :)

File: applications/CoSimulationApplication/tests/cpp_tests/test_conversion_utilities.cpp (hunk 6 above)
Comment (ashishdarekar) [reply applications/CoSimulationApplication/tests/cpp_tests/test_conversion_utilities.cpp]: True, 1e-12 okay ?

File: applications/CoSimulationApplication/tests/test_coupling_operations.py (hunk 7 above)
Comment (ashishdarekar) [reply applications/CoSimulationApplication/tests/test_coupling_operations.py]: Yes, assertVectorAlmostEqual() worked.

File: applications/CoSimulationApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 1 above)
Comment (ashishdarekar) [reply applications/CoSimulationApplication/custom_python/add_custom_utilities_to_python.cpp]: ohh ..Fine . As it is static function, hence we need to do like this. Got it. Thanks :))

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.h [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Author1 Fullname
+//                   Author2 Fullname
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.h]: missing names :)

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+//  Main authors:    Philipp Bucher
+//                   Ashish Darekar
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/parallel_utilities.h"
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp:19]:
```suggestion
#include "includes/variables.h"
#include "utilities/parallel_utilities.h"
```

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.h [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+// System includes
+#include <string>
+#include <iostream>
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/variables.h"
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.h]:
```suggestion
```

**Review by philbucher — COMMENTED** [review]
minor, rest is ok

Also I changed my mind, lets stick with `FORCE` for now and refactor in the future if necessary

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.h (hunk 8 above)
Comment (ashishdarekar) [reply applications/CoSimulationApplication/custom_utilities/conversion_utilities.h]: Added

**Comment by ashishdarekar** [comment]
@philbucher I removed hardcoded FORCE variable and made it as a input parameter. Is it okay ? or should I keep is as it ?

File: applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+#include "includes/variables.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/variable_utils.h"
+#include "utilities/atomic_utilities.h"
+#include "conversion_utilities.h"
+
+namespace Kratos {
+
+//void ConversionUtilities::ConvertElementalDataToNodalData(ModelPart& rModelPart)
+void ConversionUtilities::ConvertElementalDataToNodalData(ModelPart& rModelPart, Variable<array_1d<double,3> >& rVariable)
```
Comment (philbucher) [inline applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]: cool!
But if we make the input of the variable necessary, then we also need them for nodal and elemental var:

```suggestion
void ConversionUtilities::ConvertElementalDataToNodalData(
    ModelPart& rModelPart, 
    const Variable<array_1d<double,3> >& rElementalVariable,
    const Variable<array_1d<double,3> >& rNodalVariable)
```

Comment (ashishdarekar) [reply applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]: Ya, True, Previously I thought its just a name `FORCE`, but it is a **Variable** `FORCE`. I have made changes. 
So, While calling this, I need to use like this
`ConversionUtilities::ConvertElementalDataToNodalData(test_model_part, FORCE, FORCE);`
right ?

Comment (philbucher) [reply applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]: yes exactly

Comment (ashishdarekar) [reply applications/CoSimulationApplication/custom_utilities/conversion_utilities.cpp]: done

**Review by philbucher — APPROVED** [review]

**Comment by ashishdarekar** [comment]
Nightly Build passed !!! Merging this PR now. 
Thanks @philbucher :)

**Pull request merged**